### DISCUSSION TITLE
Per-project formatter

### DISCUSSION THREAD

**Discussion by leana8959** [mrcjkb/haskell-tools.nvim#347] [Q&A] [answered]
Hello,

I use both `stylish-haskell` and `fourmolu` to format my code. I like `stylish-haskell` for the vertical alignment and `fourmolu` for the diff-friendliness. I would like to have a way to switch between these two for projects in different contexts.

Is there an easy way to switch between the two? I have thought about creating an autocmd that toggles the option based on the project directory. The problem is that I don't know when exactly `haskell-tools.nvim` gets loaded, because my setting has to be loaded before it (right?)

Thanks for the help and have a nice day =)

**Comment by mrcjkb** [comment, marked as answer]
Hey :wave:

haskell-tools doesn't do formatting. That's done by haskell-language-server, which has plugins for the for the different formatting providers.

For configuration via this plugin, see [the advanced configuration section in the readme](https://github.com/mrcjkb/haskell-tools.nvim?tab=readme-ov-file#gear-advanced-configuration).

For haskell-language-server settings, see [the hls documentation](https://haskell-language-server.readthedocs.io/en/stable/configuration.html#configuring-haskell-language-server).

haskell-language-server loads the configuration at startup, and can't dynamically change the formatting provider at runtime.

If you only need different settings per project, you can use a `hls.json` (see [this readme section](https://github.com/mrcjkb/haskell-tools.nvim/tree/master?tab=readme-ov-file#how-to-dynamically-load-different-haskell-language-server-settings-per-project)).

To change the provider within a session, for the same project, you would have to change the configuration and then restart the LSP client with `:HlsRestart` (or `vim.cmd.HlsRestart()` in lua).

To change the configuration dynamically, you can do something like this:

```lua
vim.g.hls_formatting_provider = 'fourmolu'

vim.g.haskell_tools = {
  hls = {
    settings = function(project_root)
      local ht = require('haskell-tools')
      local defaults = ht.lsp.load_hls_settings(project_root)
      local result = vim.tbl_deep_extend('force', defaults, {
        haskell = {
          formattingProvider = vim.g.hls_formatting_provider,
        },
      })
      return result 
    end,
  },
}
```

And then define a command or keymap that sets `vim.g.hls_formatting_provider` and restarts haskell-language-server.

**Reply by leana8959** [reply]
Thank you for the help! I managed to got it work with your answer =)
